Update recent route for chars/bytes.

SirBogman opened the pull request:
Use one row per scoring so that the rank and the downward chart icon can be displayed properly.

Example:

<img width="674" alt="Screen Shot 2020-09-11 at 12 51 23 PM" src="https://user-images.githubusercontent.com/53135437/92952273-b2482680-f42d-11ea-9dc6-ed6a5e85adff.png">

SirBogman commented:
I also tested this with beta = false and using update_sql_from_api.go. It looks the same in that case.

JRaspass commented:
What about merging the scoring column into strokes? How does the following look? I think it's easier to read but I welcome your feedback, hopefully the words are roughly the same width to not be too jarring (monospace fonts tend to look ugly).
```
Fizz Buzz | Golf 2 | Python | 60 (bytes) 3rd | a min ago
Fizz Byzz | Golf 2 | Python | 60 (chars) 3rd | a min ago
...
```
I guess in most cases there will two rows per recent solution.

Irratix commented:
I'm mildly confused. Does this implementation mean that if you upload 1 solution that improves both on your bytes score and on your chars score, it shows up as 2 separate solutions on the "recent solutions" tab?

JRaspass commented:
My understanding is initially everyone's current solution becomes their best chars and bytes solution, the UI and the database then allow these solutions to diverge if you submit a solution that beats the stokes in only one criteria. But finally it would also be possible for the solutions to re-merge if you submit a solution that wins on both criteria.

SirBogman commented:
> I'm mildly confused. Does this implementation mean that if you upload 1 solution that improves both on your bytes score and on your chars score, it shows up as 2 separate solutions on the "recent solutions" tab?

That's what I was thinking.

SirBogman commented:
How about this:

<img width="683" alt="Screen Shot 2020-09-11 at 1 35 26 PM" src="https://user-images.githubusercontent.com/53135437/92955797-d3137a80-f433-11ea-89c3-9ac1731f68d3.png">

Irratix commented:
Oh yes, I think that's preferable. It makes it much easier to read what the actual score is and by what metric in just one look. I quite like this implementation.

btnlq commented:
Maybe show
```
Fizz Buzz | Golf2 | Python | 60 | Bytes | 3rd | a min ago
Fizz Buzz | Golf2 | Python | 60 | Chars | 3rd | a min ago
```
as
```
Fizz Buzz | Golf2 | Python | 60 | Both | 3rd | a min ago
```
It seems to me that most of the solutions have the same scoring and place for chars and bytes.

SirBogman commented:
> Maybe show
> 
> ```
> Fizz Buzz | Golf2 | Python | 60 | Bytes | 3rd | a min ago
> Fizz Buzz | Golf2 | Python | 60 | Chars | 3rd | a min ago
> ```
> 
> as
> 
> ```
> Fizz Buzz | Golf2 | Python | 60 | Both | 3rd | a min ago
> ```
> 
> It seems to me that most of the solutions have the same strokes and place for chars and bytes.

That's an interesting idea. I will think about it.

SirBogman commented:
At some point I will probably also add a filter for the scoring (All Scorings/Chars/Bytes). That could also help to avoid displaying duplicate rows.

SirBogman commented:
Hmmm. Something isn't quite right with this SQL. I need to debug it more before it's merged.

SirBogman commented:
I think I have a solution but I need to test it more.

SirBogman commented:
I fixed the SQL (I wasn't checking the scoring column in the join and it was resulting in extra rows on the beta site).

I also implemented @btnlq's suggestion. If two adjacent rows appear the same, other than scoring, it collapses them into a row with scoring "Both".

Here's an example where Golf1's two rows are not collapsed because the rank is different:

<img width="699" alt="Screen Shot 2020-09-11 at 11 00 29 PM" src="https://user-images.githubusercontent.com/53135437/92985746-6b851b80-f483-11ea-9bb6-ba56d0be2dd4.png">

And here's what happens when Golf1 updates their Chars solution:

<img width="706" alt="Screen Shot 2020-09-11 at 11 01 12 PM" src="https://user-images.githubusercontent.com/53135437/92985755-7b9cfb00-f483-11ea-9149-f81b01b104dd.png">

How does it look?

JRaspass commented:
I like it, on the HTML/CSS front I would be tempted to lose the Scoring header and colspan the Strokes header across the three columns (strokes, scoring, emoji) and then maybe center align the header, but I'm happy to have a play after you if you like?

JRaspass commented on routes/recent.go:
```diff
@@ -103,8 +114,20 @@ func Recent(w http.ResponseWriter, r *http.Request) {
+		} else {
+			recents = append(recents, r)
+			previous = r
+		}
```
ah i see the grouping is done at the application level rather than the sql level, I guess this means the row count could fluctuate from 25-50 depending on how many rows are collapsed. Probably not the end of the world and I presume getting the SQL to apple the merging before the limit is harder and probably fairly expensive.

SirBogman replied:
Yeah I didn't really want to think about how to do it in SQL but I'm sure it's possible. Perhaps we should double the number of rows shown so that the amount of data people see doesn't suddenly decrease?

SirBogman commented:
> I like it, on the HTML/CSS front I would be tempted to lose the Scoring header and colspan the Strokes header across the three columns (strokes, scoring, emoji) and then maybe center align the header, but I'm happy to have a play after you if you like?

Sure, play around with the header layout after merging it.

btnlq commented:
`Solution` function inserts solution code into table `code`. It then looks up this code in `code` to determine its id. But there might be multiple solutions with the same code, so it might find the wrong id. It might be worth using `RETURNING id` in `INSERT`.

This will make it possible to write the correct collapse condition:
```
previous.CodeId = r.CodeId &&
...
```
instead of
```
previous.Login == r.Login &&
previous.Hole.ID == r.Hole.ID &&
previous.Lang.ID == r.Lang.ID &&
...
```

I am also confused by the fact that `solutions` rows for the same solution may have different `submitted` times.

SirBogman commented:
> I am also confused by the fact that `solutions` rows for the same solution may have different `submitted` times.

I noticed this too and I'm working on a fix.

> If I understand correctly, `submitted`, `user_id`, `hole`, `lang` and `failing` columns can be moved to table `code` to avoid duplication (i.e. to normalize database). Then table `code` can be renamed `solutions`, and `solutions` can be renamed `scorings`. I don't know if this is possible and if this makes sense.

The `code` table is fairly new and previously there was only a `solutions` table. I think the idea of the `code` table was to essentially use less storage space by deduplicating strings. For example, if different people submitted the same code, there would be one row in the code table. My personal preference would probably be for the database to handle the deduplication of strings, if if there was enough duplication for it to be a problem.

> `Solution` function inserts solution code into table `code`. It then looks up this code in `code` to determine id. But there might be multiple solutions with the same code, so function might find the wrong id. It might be worth using `RETURNING id` in `INSERT`.

This is a good idea, but I think @JRaspass tried it and it didn't work because of `ON CONFLICT DO NOTHING`.

btnlq commented:
I'm sorry, I didn't notice `EXCLUDE USING hash (code WITH =)` constraint. This removes my previous questions.
Do I understand correctly that old long solutions are not removed from `code`? `delete_orphaned_code` trigger was removed for some reason.

JRaspass commented:
@btnlq The trigger was removed because there was a race which led to #250 

I'm working on a fix that uses locking to solve this, something like this (WIP):

```sql
CREATE FUNCTION save_solution(
    code    text,
    hole    hole,
    lang    lang,
    scoring scoring,
    user_id int
) RETURNS void AS $$
#variable_conflict use_variable
DECLARE
    bytes   int;
    chars   int;
    code_id int;
BEGIN
    bytes := octet_length(code);
    chars :=  char_length(code);

    -- Ensure we're the only one messing with code.
    LOCK TABLE code IN EXCLUSIVE MODE;

    -- Lookup the code ID, creating a new record if necessary.
    SELECT id INTO code_id FROM code WHERE code.code = code;
    IF NOT FOUND THEN
        INSERT INTO code (code) VALUES (code) RETURNING id INTO code_id;
    END IF;

    -- Update the code if it's the same length or less, but only update the
    -- submitted time if the solution is shorter. This avoids a user moving
    -- down the leaderboard by matching their personal best.
    INSERT INTO solutions (code_id, hole, lang, scoring, user_id)
         VALUES           (code_id, hole, lang, 'chars', user_id)
    ON CONFLICT ON CONSTRAINT solutions_pkey
    DO UPDATE SET failing = false,
                submitted = CASE
                    WHEN solutions.failing
                        OR chars
                        < (SELECT code.chars FROM code WHERE id = solutions.code_id)
                    THEN excluded.submitted
                    ELSE solutions.submitted
                END,
                    code_id = CASE
                    WHEN solutions.failing
                        OR chars
                        <= (SELECT code.chars FROM code WHERE id = solutions.code_id)
                    THEN excluded.code_id
                    ELSE solutions.code_id
                END;

    -- Remove any orphaned code.
    DELETE FROM code WHERE NOT EXISTS (SELECT 1 FROM solutions WHERE code_id = id);
END;
$$ LANGUAGE plpgsql;
```

The function doesn't yet handle `bytes` as well as `chars` and I intend to add some more tests but so far it's looking good. Removing the old code entries is just a nice to have though to ensure the DB doesn't get too full it shouldn't have any user facing effect either way.

@SirBogman In terms of the different times my suggestion would be for the above function to take a timestamp, this'll actually let me test the "This avoids a user moving down the leaderboard by matching their personal best." caluse. Also I'm thinking of folding the trophy logic into this function too and it would be nice to use the same timestamp for trophy unlocks.

JRaspass commented:
@SirBogman Looks like I caused a trivial conflict in the template, sorry about that, happy to merge once that's fixed though.
